File: frontend/src/services/CartService.js
import apiClient from './ApiClient';

class CartService {
  // 장바구니 항목 조회
  static async getCartItems(userId) {
    return await apiClient.get(`/cart/user/${userId}`);
  }

  // 장바구니에 상품 추가
  static async addToCart(userId, productId, quantity = 1) {
    return await apiClient.post('/cart', {
      userId,
      productId,
      quantity
    });
  }

  // 장바구니 항목 수량 업데이트
  static async updateCartItem(cartItemId, quantity) {
    return await apiClient.put(`/cart/${cartItemId}`, {
      quantity
    });
  }

  // 장바구니에서 항목 제거
  static async removeFromCart(cartItemId) {
    return await apiClient.delete(`/cart/${cartItemId}`);
  }

  // 장바구니 비우기
  static async clearCart(userId) {
    return await apiClient.delete(`/cart/user/${userId}`);
  }

  // 장바구니 항목 개수 조회
  static async getCartItemCount(userId) {
    return await apiClient.get(`/cart/user/${userId}/count`);
  }

  // 장바구니 총 금액 조회
  static async getCartTotal(userId) {
    return await apiClient.get(`/cart/user/${userId}/total`);
  }
}

export default CartService;

File: backend/src/main/java/com/shoplite/service/CartService.java
package com.shoplite.service;

import com.shoplite.model.CartItem;
import com.shoplite.model.Product;
import com.shoplite.model.User;
import com.shoplite.repository.CartItemRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class CartService {

    @Autowired
    private CartItemRepository cartItemRepository;

    @Autowired
    private UserService userService;

    @Autowired
    private ProductService productService;

    public List<CartItem> getCartItems(Long userId) {
        User user = userService.getUserById(userId);
        return cartItemRepository.findByUser(user);
    }

    public CartItem addToCart(Long userId, Long productId, Integer quantity) {
        User user = userService.getUserById(userId);
        Product product = productService.getProductById(productId);

        // 재고 확인
        if (product.getStockQuantity() < quantity) {
            throw new RuntimeException("재고가 부족합니다. 현재 재고: " + product.getStockQuantity());
        }

        // 이미 장바구니에 있는 상품인지 확인
        Optional<CartItem> existingItem = cartItemRepository.findByUserAndProduct(user, product);
        
        if (existingItem.isPresent()) {
            // 기존 항목의 수량 업데이트
            CartItem cartItem = existingItem.get();
            int newQuantity = cartItem.getQuantity() + quantity;
            
            if (product.getStockQuantity() < newQuantity) {
                throw new RuntimeException("재고가 부족합니다. 현재 재고: " + product.getStockQuantity());
            }
            
            cartItem.setQuantity(newQuantity);
            return cartItemRepository.save(cartItem);
        } else {
            // 새 항목 추가
            CartItem cartItem = new CartItem(user, product, quantity);
            return cartItemRepository.save(cartItem);
        }
    }

    public CartItem updateCartItem(Long cartItemId, Integer quantity) {
        Optional<CartItem> cartItemOpt = cartItemRepository.findById(cartItemId);
        if (cartItemOpt.isEmpty()) {
            throw new RuntimeException("장바구니 항목을 찾을 수 없습니다. ID: " + cartItemId);
        }

        CartItem cartItem = cartItemOpt.get();
        Product product = cartItem.getProduct();

        // 재고 확인
        if (product.getStockQuantity() < quantity) {
            throw new RuntimeException("재고가 부족합니다. 현재 재고: " + product.getStockQuantity());
        }

        cartItem.setQuantity(quantity);
        return cartItemRepository.save(cartItem);
    }

    public void removeFromCart(Long cartItemId) {
        Optional<CartItem> cartItemOpt = cartItemRepository.findById(cartItemId);
        if (cartItemOpt.isEmpty()) {
            throw new RuntimeException("장바구니 항목을 찾을 수 없습니다. ID: " + cartItemId);
        }

        cartItemRepository.delete(cartItemOpt.get());
    }

    public void clearCart(Long userId) {
        User user = userService.getUserById(userId);
        cartItemRepository.deleteByUser(user);
    }

    public Long getCartItemCount(Long userId) {
        User user = userService.getUserById(userId);
        return cartItemRepository.countByUser(user);
    }

    public Double getCartTotal(Long userId) {
        List<CartItem> cartItems = getCartItems(userId);
        return cartItems.stream()
                .mapToDouble(item -> item.getProduct().getPrice() * item.getQuantity())
                .sum();
    }

    public void removeFromCartByUserAndProduct(Long userId, Long productId) {
        User user = userService.getUserById(userId);
        Product product = productService.getProductById(productId);
        cartItemRepository.deleteByUserAndProduct(user, product);
    }
}



